languages/python/django-oso/django_oso/partial.py

```diff
@@ -0,0 +1,99 @@
+from django.db.models import Q
+
+from polar.expression import Expression
+from polar.variable import Variable
+from polar.exceptions import UnsupportedError
+
+
+def partial_to_query_filter(partial, type_name):
+    """
+    Convert a partial expression to a django query ``Q`` object.
+
+    Example expression structure::
+
+        Expression(And, [
+            Expression(Isa, [
+                Variable('_this'),
+                Pattern(test_app::Post, {})]),
+            Expression(Isa, [
+                Variable('_this'),
+                Pattern(test_app::Post, {})]),
+            Expression(Unify, [
+                False,
+                Expression(
+                    Dot, [
+                        Variable('_this'),
+                        'is_private'])])])
+
+    Output::
+
+        Q(is_private=False)
+    """
+
+    q = and_expr(partial, type_name)
```

Comment:
Is this always an and?

Reply:
Yes, right now. Partial constraints are interpreted conjunctively, so the natural interpretation is as an `And` expression. But when the simplifier improves we should be able to, e.g., simplify a one-argument `And` to just the arg.